patches/server/0312-Mob-Spawner-API-Enhancements.patch

```diff
@@ -101,3 +101,131 @@ index e8f888b9e37454d9863fbaf0f0c3dc982d582f1e..95b01ddddb1ba90da495927099147e77
++    }
++
++    @Override
++    public void setDelay(int delay) {
++        getHandle().getSpawner().spawnDelay = delay;
++    }
++
++    @Override
++    public int getMinSpawnDelay() {
++        return getHandle().getSpawner().minSpawnDelay;
++    }
++
++    @Override
++    public void setMinSpawnDelay(int delay) {
++        getHandle().getSpawner().minSpawnDelay = delay;
++    }
++
++    @Override
++    public int getMaxSpawnDelay() {
++        return getHandle().getSpawner().maxSpawnDelay;
++    }
++
++    @Override
++    public void setMaxSpawnDelay(int delay) {
++        getHandle().getSpawner().maxSpawnDelay = delay;
++    }
++
++    @Override
++    public int getSpawnCount() {
++        return getHandle().getSpawner().spawnCount;
++    }
++
++    @Override
++    public void setSpawnCount(int spawnCount) {
++        getHandle().getSpawner().spawnCount = spawnCount;
++    }
++
++    @Override
++    public int getMaxNearbyEntities() {
++        return getHandle().getSpawner().maxNearbyEntities;
++    }
++
++    @Override
++    public void setMaxNearbyEntities(int maxNearbyEntities) {
++        getHandle().getSpawner().maxNearbyEntities = maxNearbyEntities;
++    }
++
++    @Override
++    public int getRequiredPlayerRange() {
++        return getHandle().getSpawner().requiredPlayerRange;
++    }
++
++    @Override
++    public void setRequiredPlayerRange(int requiredPlayerRange) {
++        getHandle().getSpawner().requiredPlayerRange = requiredPlayerRange;
++    }
++
++    @Override
++    public int getSpawnRange() {
++        return getHandle().getSpawner().spawnRange;
++    }
++
++    @Override
++    public void setSpawnRange(int spawnRange) {
++        getHandle().getSpawner().spawnRange = spawnRange;
++    }
++
++    @Override
++    public boolean isActivated() {
++        return getHandle().getSpawner().isNearPlayer(getHandle().getLevel(), getHandle().blockPosition());
++    }
++
++    @Override
++    public void resetTimer() {
++        getHandle().getSpawner().delay(getHandle().getLevel(), getHandle().blockPosition());
++    }
++
++    @Override
++    public void setSpawnedItem(org.bukkit.inventory.ItemStack itemStack) {
++        com.google.common.base.Preconditions.checkArgument(itemStack != null && !itemStack.getType().isAir(),
++            "spawners cannot spawn air");
++        net.minecraft.world.item.ItemStack item = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(itemStack);
++        net.minecraft.nbt.CompoundTag entity = new net.minecraft.nbt.CompoundTag();
++        entity.putString("id", net.minecraft.core.registries.BuiltInRegistries.ENTITY_TYPE.getKey(net.minecraft.world.entity.EntityType.ITEM).toString());
++        entity.put("Item", item.save(new net.minecraft.nbt.CompoundTag()));
++        getHandle().getSpawner().setNextSpawnData(getHandle().getLevel(), getHandle().blockPosition(),
++            new net.minecraft.world.level.SpawnData(entity, java.util.Optional.empty()));
++    }
```

Comment:
Instead of adding all of these to the Craft class, make a separate interface with a bunch of default methods like PaperSpawnerHolder. Then only implement the one method you need to in both CraftMinecrartMobSpawner and CraftCreatureSpawner. [Take a look at what I did for something almost identical to this for command blocks and their minecarts](https://github.com/PaperMC/Paper/blob/d5f1ffc2e6928a6728f91ab2ad2e020b105eab7f/patches/server/0682-More-CommandBlock-API.patch).